pkg/apis/eventing/v1alpha1/cluster_provisioner_types.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"encoding/json"
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterProvisioner provisions.
+type ClusterProvisioner struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterProvisionerSpec   `json:"spec"`
+	Status ClusterProvisionerStatus `json:"status"`
+}
+
+// Check that ClusterProvisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*ClusterProvisioner)(nil)
+var _ apis.Defaultable = (*ClusterProvisioner)(nil)
+var _ runtime.Object = (*ClusterProvisioner)(nil)
+var _ webhook.GenericCRD = (*ClusterProvisioner)(nil)
+
+// ClusterProvisionerSpec is the spec for a ClusterProvisioner resource.
+type ClusterProvisionerSpec struct {
```

Comment:
It's not possible to define arguments for the Provisioner, either via `[]Parameter` or `runtime.RawExtension`. Will that be a followup?

Reply:
I think the arguments for the Provisioner are in the Channel or Source.

For 0.1, I think we should defer validation of the `runtime.RawExtension`, and perform the JSON Schema validation from Provisioner -> Channel in 0.2. If so, we can put that plan into the spec, but not implement it in CRD yet.

What do you think? Do we need the argument validation right out of the gate?

Reply:
:+1: I'm fine with delaying this until 0.2.